### PR TITLE
Vue3: Rollback v7 breaking change and keep reactive v6-compatible API

### PR DESCRIPTION
**Pull request by kasperpeulen**
In v7 an unplanned/undocumented breaking change was introduced. We apply now args as props to the component returned from the render function:

```ts
export const SomeStory = {
  render: (args) => ({
    components: { TestBtn },
    setup() {
      return { args };
    },
    template: `
    <div>
      <TestBtn v-bind="args"></TestBtn>
    </div>`,
  }),
  argTypes: {
    onClick: { action: 'clicked' },
  },
  args: {
    disabled: false,
  }
};
```

In this example, that means that the onClick action and the disabled arg are not only applied to `TestBtn`, but also to the outer div. This is because of fall through attributes:
https://vuejs.org/guide/components/attrs.html#fallthrough-attributes

```ts
// simplistic Storyook vue3 implementation
createApp({
  setup() {
    // modify args in the storyContext to make them reactive
    storyContext.args = reactive(storyContext.args);
    // apply args to user provided render function
    const Component = SomeStory.render(storyContext.args);
    return () => {
      // Behavior in v7: Apply args also as props to component options returned from the story render
      return h(Component, storyContext.args);
      // Behavior in v6: Just render the Component without props, args are already provided in the closure.
      return h(Component);
    };
  }
})
```

This breaking change has as benefit that the user doesn't have to write v-bind="args" here:
`template: '<MyComponent v-bind="args" />',`
They can leave out `v-bind="args"`, but only if the component is on the root.

The major downside of this change is if you have something else than your component on the root:
`template: '<div><MyComponent v-bind="args" /></div>',`
Now, the args are "magically" spread into a div, but the user didn't write the args for the div, they are written and meant for `MyComponent`. We are getting feedback from multiple Storybook users that people don't expect this, causing unexpected issues.

I think the reason that people don't expect this isn't because they don't understand Vue. It's more that they don't expect the template that they give to Storybook to be invoked as a component with their args as props. The user cannot see how SB will invoke the component, and args are already available in the closure. It is not intuitive that args are also available via a second route.

This is different from all our other renderers (or at least the ones I know):
React: `render: (args) => <Component {...args} />`
Svelte: `render: (args) => { Component, props: args }`
Html: `render: (args) => createButton(args)`
Lit: `render: (args) => <lit-element label=${args.label}></lit-element>`
Args are just available in the closure and the element returned from render is not invoked a second time.

Also, our TS types and the add-on controls will match the shape of the component, not the root tag of the template. So even our tools don't expect this, and you will get warnings from TS if you would write an arg that is not compatible with the component (even if it would be compatible with the root tag/component).

That being said, @chakAs3 is going to write up a RFC for doing this breaking change behind a feature flag in V7 and if this gains popularity, we can enable this feature by default for v8!

### CODE REVIEW

**Comment by chakAs3** [comment]
> I think the reason that people don't expect this isn't because they don't understand Vue. It's more that they don't expect the template that they give to Storybook to be invoked as a component with their args as props. The user cannot see how SB will invoke the component

i completely agree with you, and moreover they don't understand how Storybook works.  we need to work on this to make it clear for newbie and experts.  better documentation and code examples.

File: code/renderers/vue3/src/render.ts [hunk 1]
```diff
@@ -161,11 +163,12 @@ function teardown(
...
-    if (app?.reactiveSlots) updateArgs(app.reactiveSlots, slots);
-    return app?.reactiveSlots;
-  }
+  // this seem to cause recursive updates, and vue errors
+  // if (slotsMap.has(storyID)) {
+  //   const app = slotsMap.get(storyID);
+  //   if (app?.reactiveSlots) updateArgs(app.reactiveSlots, slots);
+  //   return app?.reactiveSlots;
+  // }
   slotsMap.set(storyID, { component, reactiveSlots: slots });
```
Comment (kasperpeulen) [inline code/renderers/vue3/src/render.ts:172]: @chakAs3  I think I can remove the whole `slotsMap`. All the reactive test passes without it.
My gut feeling is that line 25, fixes this. 
Without it, the slots are basically static, and need to updated forcefully.
But now we fix the render function, we don't need this code anymore.

**Comment by yannbf** [comment]
I have tested different reactivity scenarios and they all seem to work fine in this PR, however the build-sandboxes step is failing and preventing the PR to proceed with the tests on the sandboxes.

https://github.com/storybookjs/storybook/pull/22259 has to be fixed first, then merged into this branch, so we can get a proper look at the CI results.

**Comment by chakAs3** [comment]
> We are getting feedback from multiple Storybook users that people don't expect this, causing unexpected issues.

Hi Kasper I'm quite available on Discord , and github i'm not getting any feedback apart from the one that you forward to me  and even these guys are not coming back to us although all this their use case,  if there is any other channel that you use to receive feedback please share with me, i can't tackle vue issues.

**Comment by chakAs3** [comment]
> This is different from all our other renderers (or at least the ones I know): React: `render: (args) => <Component {...args} />` Svelte: `render: (args) => { Component, props: args }` Html: `render: (args) => createButton(args)` Lit: `render: (args) => <lit-element label=${args.label}></lit-element>` Args are just available in the closure and the element returned from render is not invoked a second time.

Here to be honest i don't see any difference in Storybook api,  render function return a Renderer Component.
the difference that you don't seem to accept is that Vue , React, Svelte, ... are by definition everyone has its api and rendering mecanism.
what i see here is the opposite, you are traiting Vue renderer with kind of discrimination by saying i don't like your fallthough  attribute feature.

**Comment by chakAs3** [comment]
> Also, our TS types and the add-on controls will match the shape of the component, not the root tag of the template. So even our tools don't expect this, and you will get warnings from TS if you would write an arg that is not compatible with the component (even if it would be compatible with the root tag/component).

i feel like i don't know Storybook really with this statement, this not about Vue, but How Storybook real works .
please correct me if i'm wrong.

Component Story in its default state, is to document different state of a specific Component.
so basic and simple form is 
```ts
const meta = {
 
  component: Button,

} satisfies Meta<typeof Button>;

export const Primary: Story = {
  args: {
    primary: true,
    label: 'Button',
  },
};
```

but this is just basic form, does not mean my args should match the component's  props like you state @kasperpeulen 
so i'm not sure if you know that we can do something like this 
```ts 
export const CustomStory: Story = {
  args: { 
    customStoryArg: 'customStoryArg',
  },
  render: (args: any) => ({
    setup() {
      return { args };
    },
    template: '<div> I dont need the Button in this Story, Hello some custom args {{ args.customStoryArg }} </div>',
  }),
};
```
it will just work 
image

you still can document your Component, and you can have you custom story,  when you custom your render and go from the default one which the component, You really custom your render you are not attached the component here it does not matter , like in your statement 
`
The major downside of this change is if you have something else than your component on the root:
template: '<div><MyComponent v-bind="args" /></div>',`

it does not really matter your template could be '<div></div><MyComponent v-bind="args" />' or <div></div> or anything here you creating your custom rendering component you just have to be good at this Renderer ( Vue, Svelte, React ) and know exactly how it works you life will be fine, the reability shifts the the renderer here we don't talk about Storybook API, this part is delegate to the renderer Storybook should not be altering the behavior here, it is very dangerous and we should not forget one important thing the End User aims to test  develop his UI Library(vue/svelte) , he needs full features

**Comment by chakAs3** [comment]
> > Also, our TS types and the add-on controls will match the shape of the component, not the root tag of the template. So even our tools don't expect this, and you will get warnings from TS if you would write an arg that is not compatible with the component (even if it would be compatible with the root tag/component).
> 
> it is a bold statement, i feel like i don't know Storybook really with this statement, this is not about Vue this time, but about How Storybook real works . @shilman @yannbf @tmeasday . Please correct me if i'm wrong.
> 
> Component Story in its default state, is to document different state of a specific Component. so basic and simple form is
> 
> ```ts
> const meta = {
>  
>   component: Button,
> 
> } satisfies Meta<typeof Button>;
> 
> export const Primary: Story = {
>   args: {
>     primary: true,
>     label: 'Button',
>   },
> };
> ```
> 
> 
> It's important to note that the basic form of a Component Story does not necessarily require the `args` to match the component's props. As @kasperpeulen stated, the default purpose of a Component Story is to document different states of a component. Therefore, it's possible to use the `args` to specify any relevant information for the particular story being documented, even if it doesn't directly relate to the component's props.
> 
> For example, you can define a `CustomStory` that doesn't need the `Button` component and uses custom arguments instead. The following code shows an example of such a `CustomStory`:
> 
> ```ts
> export const CustomStory: Story = {
>   args: { 
>     customStoryArg: 'customStoryArg',
>   },
>   render: (args: any) => ({
>     setup() {
>       return { args };
>     },
>     template: '<div> I don\'t need the Button in this Story. Hello, some custom args: {{ args.customStoryArg }} </div>',
>   }),
> };
> ```
> 
> In this example, the `CustomStory` defines a custom `customStoryArg` argument that is used in the template. The `render` function uses the `setup()` function to return the `args` object, which is then passed to the template as a property. This way, the `CustomStory` is documenting a custom use case that doesn't require the `Button` component.
> 
> image
> 
> you still can document your Component, and you can have you custom story, when you custom your render and go from the default one which the component, You really custom your render you are not attached the component here it does not matter , like in your statement ` The major downside of this change is if you have something else than your component on the root: template: '<div><MyComponent v-bind="args" /></div>',`
> 
> Regardless of the template you use to create your custom rendering component - whether it's
> 
> or
> or something else entirely - it's important to have a good understanding of the Vue renderer and how it works. The Vue renderer is responsible for rendering your component, and its reliability determines the reliability of your component. Therefore, it's important to rely on the Vue renderer for reliability and not on Storybook API. Any interference by Storybook API with the behavior of the Vue renderer could be risky and should be avoided. Finally, it's crucial to keep in mind that the end user developing their UI library (Vue/Svelte) needs full access to the features provided by Vue to effectively test and develop their library.

**Review by shilman — APPROVED** [review]
@kasperpeulen

Thanks so much for this. We didn't realize the new API had been introduced in v7. @chakAs3 and I will write an RFC for why it's important and more Vue-like to automatically pass args as props to stories, and we will seriously consider updating the story API to make it more idiomatic for Vue users.

@chakAs3 

If you consider the new API as a feature, removing it is a breaking change and should only happen in a major release with proper deprecation, etc.

However, the new API was not documented until Kasper wrote tests for it, and is still not documented--our official docs for Vue3 all recommend using the setup function (the v6 API). So it's more like an accidental change that snuck in. Removing it is still breaking for anybody who uses it, and it sucks. But it sucks less than committing ourselves to an API that we are not yet comfortable with.

I appreciate your passion around the alternative, your patience to discuss this via an RFC, and of course all of the hard work you're doing to make Storybook the best way to develop, document, and test Vue components in isolation.

**Comment by chakAs3** [comment]
@shilman  @kasperpeulen  i got your point maybe we did not do think properly, I just wanna mention this may not have with specific issue but it is a really good opportunity to be more explicit, and document our API before any release, I guess even with Old v6 Vue developers were confused as they don't find much content for vue, it was more about react, we will put more effort, @kasperpeulen, @jonniebigodes, @kylegach, @Integrayshaun and myself I will provide necessary info and examples.

**Comment by chakAs3** [comment]
Here is the problem I saw that, and the guy who proposed the implementation was giving this as a solution, he said clearly in His PR I'm not sure if the Vue community will agree with that, unfortunately, that was no one to veille on Vue implementation, and no one has commented this so you went ahead and merged this, and here we the real breaking change happened for me. 
his solution was very clever but was limited. in the first place, it was just to migrate from Vue2 to Vue3 format.

My RFC will not be the PR for fallthrough attributes but more exposing and normalizing  Storybook API.

**Comment by chakAs3** [comment]
> our official docs for Vue3 all recommend using the setup function (the v6 API)

Here is the problem I saw that, and the guy who proposed the implementation was giving this as a solution, he said clearly in His PR I'm not sure if the Vue community will agree with that, unfortunately, that was no one to veille on Vue implementation, and no one has commented this so you went ahead and merged this, and here we the real breaking change happened for me.
his solution was very clever but was limited. in the first place, it was just to migrate from Vue2 to Vue3 format.

My RFC will not be the PR for fallthrough attributes but more exposing and normalizing Storybook API.

**Comment by MineDrum** [comment]
Just to chime in. I'm very glad to see this getting rolled back for now. All of a sudden when I upgraded to v7 my tests were failing because when I applied a 'placeholder' prop to my components and tried to get the element by placeholder text in my test, storybook was finding two elements instead of one. This is because I originally had a decorator div where I would apply spacing and dark tailwind styling. The placeholder prop was being applied to the div and to the component which made no sense and forced me to remove any decorators that I had.

@chakAs3

**Comment by SimmeNilsson** [comment]
Also glad to see it rolled back.
For me it was causing issues with when wanting to use things like v-model="myRef" and then suddenly :myRef="myRef" was added "for free".
https://github.com/storybookjs/storybook/discussions/22156

@chakAs3

**Pull request merged**